source/grammar/qasm3.g4

```diff
@@ -1,98 +1,461 @@
+    : globalStatement
+    | expressionStatement
+    | declarationStatement
+    | branchingStatement
+    | loopStatement
+    | controlDirectiveStatement
+    | aliasStatement
+    | quantumStatement
+    | timingBox
+    | pragma
+    | comment
+    ;
+
+globalStatement: subroutineDefinition
+    | kernelDeclaration
+    | quantumGateDefinition
+    | calibrationDefinition
+    ;
+
+declarationStatement
+    : ( quantumDeclaration | classicalDeclaration | constantDeclaration) SEMICOLON
+    ;
+
+comment
+    : '//' AnyString '\n'
+    | '/*' AnyString '*/'
+    ;
+
+returnSignature
+    : ARROW classicalDeclaration
+    ;
+
+programBlock
+    : LBRACE ( programBlock | statement ) RBRACE
+    ;
+
+/* Types and Casting */
+
+designator
+    : LBRACKET expression RBRACKET
+    ;
+
+doubleDesignator
+    : LBRACKET expression COMMA expression RBRACKET
+    ;
+
+identifierList
+    : ( Identifier COMMA )* Identifier
+    ;
+
+indexIdentifier
+    : Identifier ( designator )?
+    ;
+
+indexIdentifierList
+    : ( indexIdentifier COMMA )* indexIdentifier
+    ;
+
+association
+    : COLON Identifier
+    ;
+
+// Quantum Types
+quantumType
+    : 'qubit'
+    | 'qreg'
+    ;
+
+quantumDeclaration
+    : quantumType Identifier designator
+    ;
+
+quantumArgument
+    : quantumType designator? association
+    ;
+
+quantumArgumentList
+    : ( quantumArgument COMMA )* quantumArgument
+    ;
+
+// Classical Types and Timing Types
+bitType
+    : 'bit'
+    | 'creg'
+    ;
+
+singleDesignatorType
+    : 'int'
+    | 'uint'
+    | 'float'
+    | 'angle'
+    ;
+
+doubleDesignatorType
+    : 'fixed'
+    ;
+
+noDesignatorType
+    : 'bool'
+    | timingType
+    ;
+
+classicalType
+    : singleDesignatorType designator?
+    | doubleDesignatorType doubleDesignator?
+    | noDesignatorType
+    | bitType designator?
+    ;
+
+timingType
+    : 'length'
+    | 'stretch' Integer?
+    ;
+
+constantDeclaration
+    : 'const' Identifier ASSIGN expression
+    ;
+
+singleDesignatorDeclaration
+    : singleDesignatorType designator Identifier
+    ;
+
+doubleDesignatorDeclaration
+    : doubleDesignatorType doubleDesignator Identifier
+    ;
+
+noDesignatorDeclaration
+    : noDesignatorType Identifier
+    ;
+
+classicalVariableDeclaration
+    : singleDesignatorDeclaration
+    | doubleDesignatorDeclaration
+    | noDesignatorDeclaration
+    ;
+
+classicalDeclaration
+    : classicalVariableDeclaration assignmentExpression?
+    ;
+
+classicalTypeList
+    : ( classicalType COMMA )* classicalType
+    ;
+
+classicalArgument
+    : classicalType association
+    ;
+
+classicalArgumentList
+    : ( classicalArgument COMMA )* classicalArgument
+    ;
+
+// Aliasing
+aliasStatement
+    : 'let' Identifier ASSIGN concatenateExpression
+    ;
+
+// Register Concatenation and Slicing
+concatenateExpression
+    : ASSIGN
+    ( Identifier range
+    | Identifier '||' Identifier
+    | Identifier LBRACKET expressionList RBRACKET )
+    ;
+
+range
```

Comment:
it's necessary change this "range" identifier to don't collide with Python range.
- error: symbol "range" conflicts with generated code in target language or runtime

Reply:
Addressed in https://github.com/Qiskit/openqasm/pull/89/commits/c3b3b28409aec991623c590a5ecfe28c7527d14c.